daemon/daemon.go

```diff
@@ -529,33 +532,13 @@ func (d *Daemon) initStandbyHandling() {
 
 // Start the Daemon
 func (d *Daemon) Start() {
-	// die when asked to restart (systemd should get us back up!)
-	d.overlord.SetRestartHandler(func(t state.RestartType) {
-		switch t {
-		case state.RestartDaemon:
-			d.tomb.Kill(nil)
-		case state.RestartSystem:
-			// try to schedule a fallback slow reboot already here
-			// in case we get stuck shutting down
-			if err := reboot(rebootWaitTimeout); err != nil {
-				logger.Noticef("%s", err)
-			}
-
-			d.mu.Lock()
-			defer d.mu.Unlock()
-			// remember we need to restart the system
-			d.restartSystem = true
-			d.tomb.Kill(nil)
-		case state.RestartSocket:
-			d.mu.Lock()
-			defer d.mu.Unlock()
-			d.restartSocket = true
-			d.tomb.Kill(nil)
-		default:
-			logger.Noticef("internal error: restart handler called with unknown restart type: %v", t)
-			d.tomb.Kill(nil)
-		}
-	})
+	if d.overlord == nil {
+		// we need to schedule and wait for a system restart
+		d.tomb.Kill(nil)
+		// avoid systemd killing us again while we wait
+		systemdSdNotify("READY=1")
```

Comment:
Perhaps we should use `EXTEND_TIMEOUT_USEC=...` to tell systemd we are not ready but we are not quite done starting yet. This is further documented on `sd_notify(3)`